File: master-node-docker/sentinel/dev/free.py
# coding=utf-8
import json

import falcon

from ..config import DECIMALS
from ..db import db
from ..helpers import eth_helper


def check_free(to_addr):
    tx = db.free.find_one({
        'to_addr': to_addr
    })

    return (tx is None) is False


def insert_free(to_addr):
    _ = db.free.insert_one({
        'to_addr': to_addr
    })


class GetFreeAmount(object):
    def on_post(self, req, resp):
        account_addr = str(req.body['account_addr']).lower()
        eths = int(0.25 * (10 ** 18))
        sents = int(1000 * DECIMALS)

        tx_done = check_free(account_addr)
        if tx_done is True:
            message = {
                'success': False,
                'message': 'Test Tokens already claimed'
            }
        else:
            errors, tx_hashes = eth_helper.free(account_addr, eths, sents)

            if len(errors) > 0:
                message = {
                    'success': False,
                    'errors': errors,
                    'tx_hashes': tx_hashes,
                    'message': 'Error occurred while transferring free amount.'
                }
            else:
                insert_free(account_addr)
                message = {
                    'success': True,
                    'errors': errors,
                    'tx_hashes': tx_hashes,
                    'message': 'Successfully transferred Test Tokens'
                }

        resp.status = falcon.HTTP_200
        resp.body = json.dumps(message)

File: master-node-nodejs/src/dev/free.js
import ETHHelper from '../helpers/eth'
import { DECIMALS } from '../config/vars';
import { Free } from '../models';
import database from '../db/database';

let db = null;

const checkFree = (toAddr, cb) => {
  Free.findOne({ to_addr: toAddr }, (err, tx) => {
    if (!tx) cb(false)
    else cb(true)
  })
}

const insertFree = (toAddr, cb) => {
  let data = {
    to_addr: toAddr
  }
  let freeData = new Free(data)
  database.insert(freeData, (err, resp) => {
    cb(err, resp)
  })
}

const getFreeAmount = (req, res) => {

  let accountAddr = req.body['account_addr'];
  let eths = parseInt(0.25 * Math.pow(10, 18))
  let sents = parseInt(1000 * (DECIMALS * 1.0))

  checkFree(accountAddr, (txDone) => {
    if (txDone) {
      res.send({
        'success': false,
        'message': 'Test Tokens already claimed'
      })
    } else {
      ETHHelper.free(accountAddr, eths, sents, (errors, txHashes) => {
        if (errors.length > 0) {
          res.send({
            'success': false,
            'errors': errors,
            'txHashes': txHashes,
            'message': 'Error occurred while transferring free amount.'
          })
        } else {
          insertFree(accountAddr, (err, resp) => {
            res.send({
              'success': true,
              'errors': errors,
              'txHashes': txHashes,
              'message': 'Successfully transferred Test Tokens'
            })
          });
        }
      })
    }
  })
}

export default {
  getFreeAmount
}